kits/arms/ex_impedance_control.m

```diff
@@ -0,0 +1,213 @@
+%
+% Author:        Dave Rollinson
+%                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      
+% Date:          Oct 2018
+
+% Copyright 2017-2018 HEBI Robotics
+
+%% Setup
+clear *;
+close all;
+
+HebiLookup.initialize();
+
+localDir = fileparts(mfilename('fullpath'));
+
+armName = '6-DoF + gripper';
+armFamily = 'Arm';
+
+[ armGroup, armKin, armParams ] = setupArm( armName, armFamily );
+
+% Increase feedback frequency since we're calculating velocities at the
+% high level for damping.  Going faster can help reduce a little bit of
+% jitter for fast motions, but going slower (100 Hz) also works just fine
+% for most applications.
+armGroup.setFeedbackFrequency(200); 
+
+% Double the effort gains from their default values, to make the arm more
+% sensitive for tracking force.
+gains = armGroup.getGains();
+gains.effortKp = 2 * gains.effortKp;
+armGroup.send('gains',gains);
+
+gravityVec = armParams.gravityVec;
+effortOffset = armParams.effortOffset;
+
+numDoF = armKin.getNumDoF;
+
+enableLogging = true;
+
+% Start background logging 
+if enableLogging
+   logFile = armGroup.startLog('dir',[localDir '/logs']); 
+end
+
+%% Gravity compensated mode
+cmd = CommandStruct();
+
+% Keyboard input
+kb = HebiKeyboard();
+keys = read(kb);
+
+disp('Commanded gravity-compensated zero force to the arm.');
+disp('  SPACE - Toggles an impedance controller on/off:');
+disp('          ON  - Apply controller based on current position');
+disp('          OFF - Go back to gravity-compensated mode');
+disp('  ESC - Exits the demo.');
+
+% Impendance Control Gains
+% NOTE: The gains corespond to:
+% [ trans_x trans_y trans_z rot_x rot_y rot_z ]
+%
+% Translations and Rotations can be specified in the
+% base frame or in the end effector frame.  See code below for
+% details.
+%
+% UNCOMMENT THE GAINS YOU WANT TO USE FOR A GIVEN RUN, AND COMMENT OUT ALL
+% THE OTHER GAINS.
+
+%     % HOLD POSITION ONLY (Allow rotation around end-effector position)
+%     gainsInEndEffectorFrame = true;
+%     damperGains = [5; 5; 5; .0; .0; .0;]; % (N/(m/sec)) or (Nm/(rad/sec))
+%     springGains = [500; 500; 500; 0; 0; 0];  % (N/m) or (Nm/rad)
+
+%     % HOLD ROTATION ONLY
+%     gainsInEndEffectorFrame = true;
+%     damperGains = [0; 0; 0; .1; .1; .1;]; % (N/(m/sec)) or (Nm/(rad/sec))
+%     springGains = [0; 0; 0; 10; 10; 10];  % (N/m) or (Nm/rad)
+ 
+    % HOLD POSITION AND ROTATION - BUT ALLOW MOTION ALONG/AROUND Z-AXIS
+    gainsInEndEffectorFrame = true;
+    damperGains = [10; 10; 0; .1; .1; .1;]; % (N/(m/sec)) or (Nm/(rad/sec))
```

Comment:
Really nice demo example, I like it!

The `POSITION + ROTATION` example went unstable for me several times. I did change my `X8-3 Wrist1` effort gains to be appropriately scaled from the defaults.